src/controllers/v1/_common/demand.ts

```diff
@@ -82,16 +83,48 @@ export class DemandController extends Controller {
     if (updated_since) {
       query.push(['updated_at', '>', parseDateParam(updated_since)])
     }
+    const selfAddress = (await this.addressResolver.determineSelfAddress()).address
+    const roles = env.ROLES
+    // if roles does not include admin or optimiser filter demands
+    const isPrivileged = roles.includes('optimiser') || roles.includes('admin')
+    const demands: DemandRow[] = []
+
+    if (isPrivileged) {
+      // Privileged roles get all demands matching base query
+      demands.push(...(await this.db.get('demand', query)))
+    } else {
+      // Owner's own demands
+      const ownedQuery: Where<'demand'> = [...query, ['owner', '=', selfAddress]]
+      demands.push(...(await this.db.get('demand', ownedQuery)))
+
+      // roles that contain only member_a or member_b or both
+      for (const member of roles) {
+        const demand_a_boolean = this.dbDemandSubtype === 'demand_a'
+
+        // Get all match2s where the member is either member_a or member_b
+        const match2Query: Where<'match2'> = [[member === 'member-a' ? 'member_a' : 'member_b', '=', selfAddress]]
+        const matches = await this.db.get('match2', match2Query)
+        // From the match2s get all the demands that are type this.dbDemandSubtype
+        const demandIds = matches.map((m) => (demand_a_boolean ? m.demand_a_id : m.demand_b_id))
+        if (demandIds.length > 0) {
+          const matchedQuery: Where<'demand'> = [...query, ['id', 'IN', demandIds]]
+          demands.push(...(await this.db.get('demand', matchedQuery)))
+        }
+      }
+    }
```

Comment:
This could use a refactor to make it a little easier to understand. General rules; try to avoid using `else` and try to avoid multiple levels of nesting. To avoid using `else` use early returns in the if